src/lepy/lego.go

```diff
@@ -0,0 +1,178 @@
+	URL       string `json:"url"`
+	Domain    string `json:"domain"`
+}
+
+//export RunLegoCommand
+func RunLegoCommand(message *C.char) *C.char {
+	CLIArgs, err := extractArguments(C.GoString(message))
+	if err != nil {
+		return C.CString(fmt.Sprint("error: couldn't extract arguments: ", err))
+	}
+	for k, v := range CLIArgs.Env {
+		os.Setenv(k, v)
+	}
+	certificate, err := requestCertificate(CLIArgs.Email, CLIArgs.Server, CLIArgs.CSR, CLIArgs.Plugin)
+	if err != nil {
+		return C.CString(fmt.Sprint("error: couldn't request certificate: ", err))
+	}
+	response_json, err := json.Marshal(certificate)
+	if err != nil {
+		return C.CString(fmt.Sprint("error: coudn't build response message: ", err))
+	}
+	return_message_ptr := C.CString(string(response_json))
+	return return_message_ptr
+}
+
+func requestCertificate(email, server, csr, plugin string) (*LegoOutputResponse, error) {
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't generate priv key: %s", err)
+	}
+	user := LetsEncryptUser{
+		Email: email,
+		key:   privateKey,
+	}
+	config := lego.NewConfig(&user)
+
+	// This CA URL is configured for a local dev instance of Boulder running in Docker in a VM.
+	config.CADirURL = server
+	config.Certificate.KeyType = certcrypto.RSA2048
+
+	// A client facilitates communication with the CA server.
+	client, err := lego.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create lego client: %s", err)
+	}
+
+	err = configureClientChallenges(client, plugin)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't configure client challenges: %s", err)
+	}
+
+	reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
+	if err != nil {
+		return nil, fmt.Errorf("couldn't register user: %s", err)
+	}
+	user.Registration = reg
+
+	block, _ := pem.Decode([]byte(csr))
+	if block == nil || block.Type != "CERTIFICATE REQUEST" {
+		return nil, errors.New("failed to decode PEM block containing certificate request")
+	}
+	csrObject, err := x509.ParseCertificateRequest(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse certificate request: %s", err)
+	}
+	request := certificate.ObtainForCSRRequest{
+		CSR:    csrObject,
+		Bundle: true,
+	}
+	certificates, err := client.Certificate.ObtainForCSR(request)
+	if err != nil {
+		return nil, fmt.Errorf("coudn't obtain cert: %s", err)
+	}
+
+	return &LegoOutputResponse{
+		CSR:               string(certificates.CSR),
+		PrivateKey:        string(certificates.PrivateKey),
+		Certificate:       string(certificates.Certificate),
+		IssuerCertificate: string(certificates.IssuerCertificate),
+		Metadata: Metadata{
+			StableURL: certificates.CertStableURL,
+			URL:       certificates.CertURL,
+			Domain:    certificates.Domain,
+		},
+	}, nil
+}
+
+func configureClientChallenges(client *lego.Client, plugin string) error {
+	switch plugin {
+	case "http":
```

Comment:
I'm not sure what to call this plugin.

Reply:
Maybe we can use the empty string here? So if someone does not specify a plugin for DNS01, it uses HTTP01?

Reply:
done. It's just a tiny bit annoying because now we can't have the env variable as the last argument :D.